### ISSUE TITLE
Unable to create Git repositories

### ISSUE THREAD

**Issue by tksrc** [pulumi/pulumi-azuredevops#66]
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

I am unable to create a new repository using the Git resource and initialization type of `Clean`or `Uninitialized`.

### Steps to reproduce
```python
import pulumi_azuredevops as ado

project = ado.Project(
    "testproject",
    description="Test Project Description",
    version_control="Git",
    visibility="private",
    work_item_template="Agile"
)

repo = ado.Git(
    "repo",
    project_id=project.id,
    initialization=ado.GitInitializationArgs(
        init_type="Clean",
    ),
)
```

Expected: A new repository "repo" to be created.
Actual: Pulumi errors out:
```
azuredevops:index:Git (repo):
    error: azuredevops:index/git:Git resource 'repo' has a problem: "initialization.0.service_connection_id": all of `initialization.0.service_connection_id,initialization.0.source_type,initialization.0.source_url` must be specified
```

If I specify a source_url of another repo and change the init_type to `Clone` it all works fine.

**Comment by tksrc** [comment]
The issue above is present when using pulumi-azuredevops `v2.2.0`. I just downgraded to `v2.1.1` and the issue is no longer there.

**Comment by fdebuire** [comment]
Got into the same issue with the version `v2.3.0`.

**Comment by ericogr** [comment]
Same issue with v2.3.1.

**Workaround**: you can try to add dummy values for required fields (ServiceConnectionId, SourceType and SourceUrl) just to bypass the validation, like this:

	repository, err := azuredevops.NewGit(ctx, repoName, &azuredevops.GitArgs{
		Name:          pulumi.String(repoName),
		ProjectId:     pulumi.String(project.Id),
		DefaultBranch: pulumi.Sprintf("refs/heads/main"),
		Initialization: azuredevops.GitInitializationArgs{
			InitType: pulumi.String("Clean"),
			ServiceConnectionId: pulumi.String(""),
			SourceType:          pulumi.String("Git"),
			SourceUrl:           pulumi.String("https://repo.com"),
		},
	})

>Note: if you already have a git repo using provider previous version (2.1.1), Pulumi will try to delete and create it again after this change.
(Go SDK, not tested in .NET)

**Comment by stack72** [comment]
This is related to https://github.com/pulumi/pulumi-terraform-bridge/issues/540

**Comment by lukehoban** [comment]
I believe there are two paths to addressing this:
1. Implement a complete fix for https://github.com/pulumi/pulumi-terraform-bridge/issues/540, which we likely need to be very careful with as this may have subtle interactions with other providers.
2. Fork the upstream to remove the `default` values for two of the affected properties - which is presumably what is causing trouble here (the combination of `default` and `required_with`).  https://github.com/microsoft/terraform-provider-azuredevops/blob/81649c07e6bc69b2f4d2f0a0252a33246d7bdab0/azuredevops/internal/service/git/resource_git_repository.go#L128C25-L128C25

**Comment by VenelinMartinov** [comment]
I looked into this and compared what we do here with what terraform does.

TLDR: I don't think this is a problem with us handling `RequiredWith` but rather with the provider using a default of an empty string to mean null. TF seems to handle this by doing config validation before actually setting the defaults.


This is my understanding of the problem:

The provider sets a default value for `service_connection_id` of an empty string but it kind of means null there, since the empty string is not a valid value for this.  
Terraform seems to do the [config validation](https://github.com/hashicorp/terraform/blob/13e26b60dd545533efab76ab6f48ce20b1e47f37/internal/terraform/node_resource_abstract_instance.go#L784) BEFORE [getting the defaults](https://github.com/hashicorp/terraform/blob/13e26b60dd545533efab76ab6f48ce20b1e47f37/internal/terraform/node_resource_abstract_instance.go#L846) from the provider. This means that the validation passes there since the value for `service_connection_id` is still null. Terraform then gets the default from the provider (an empty string) and then passes it to the planning phase.  
That then yields some errors (`.initialization[0].service_connection_id: planned value cty.StringVal("") for a non-computed attribute`) but they [tolerate them](https://github.com/hashicorp/terraform/blob/13e26b60dd545533efab76ab6f48ce20b1e47f37/internal/terraform/node_resource_abstract_instance.go#L908) because this is provider is using a `LegacyTypeSystem`.

In tf-bridge we seem to do the reverse - we set the defaults from the schema and only then pass the config for validation which yields the error about the `RequiredWith`.
I tested moving around the validation in terraform and it does yield the same error if I first apply the default to the config.
I also tested ignoring the error in tf-bridge and everything works fine. The provider seems to handle the empty string values which actually mean null fine.

Should we re-order the operations here? Alternatively we could add some condition about ignoring this type of errors in `ValidateResource` but I'm not sure how to make certain this doesn't yield any false negatives.

**Comment by VenelinMartinov** [comment]
We just released https://github.com/pulumi/pulumi-azuredevops/releases/tag/v2.14.1 which contains a fix for this issue.
